### PR TITLE
kubelet: use newtimer instead in nodeshutdown manager

### PR DESCRIPTION
**Pull request by jonyhy96**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

A time.After can't be cancelled so it has to run the full time specified. Besides, we should use `m.clock` as timer for better testability.

### CODE REVIEW

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 1]
```diff
@@ -326,15 +326,19 @@ func (m *managerImpl) processShutdownEvent() error {
...
 		go func() {
-			defer close(c)
+			defer close(doneCh)
 			wg.Wait()
 		}()
 
 		select {
-		case <-c:
-		case <-time.After(time.Duration(group.ShutdownGracePeriodSeconds) * time.Second):
+		case <-doneCh:
```
Comment (wzshiming) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:340]:
```suggestion
		case <-doneCh:
			timer.Stop()
```

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 2]
```diff
@@ -326,15 +326,19 @@ func (m *managerImpl) processShutdownEvent() error {
 			}(pod, group)
 		}
 
-		c := make(chan struct{})
+		var (
+			doneCh = make(chan struct{})
+			timer  = time.NewTimer(time.Duration(group.ShutdownGracePeriodSeconds) * time.Second)
+		)
+		defer timer.Stop()
```
Comment (wzshiming) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:333]:
```suggestion
```

**Review by wzshiming — CHANGES_REQUESTED** [review]
I'm guessing you want to actively release the timer in a group of stop task completions, but using `defer` in a loop probably won't do that

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 1 above)
Comment (jonyhy96) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:339]: Done

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go (hunk 2 above)
Comment (jonyhy96) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go]: Done

**Comment by manugupt1** [comment]
TIL, thanks! would it be possible to add some unit tests as well?

**Comment by jonyhy96** [comment]
> TIL, thanks! would it be possible to add some unit tests as well?

That sounds great, will issue another pr on that, still need some knowledge about how to `mock time`

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go [hunk 3]
```diff
@@ -326,15 +326,19 @@ func (m *managerImpl) processShutdownEvent() error {
...
+			defer close(doneCh)
 			wg.Wait()
 		}()
 
 		select {
-		case <-c:
-		case <-time.After(time.Duration(group.ShutdownGracePeriodSeconds) * time.Second):
+		case <-doneCh:
+			timer.Stop()
+		case <-timer.C:
```
Comment (wgahnagl) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:341]: timer.C is a little confusing, would it be possible to change this to timer.After, from here in the docs? https://pkg.go.dev/time

Comment (jonyhy96) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go:341]: A time.After can't be cancelled so it has to run the full time specified.

**Comment by manugupt1** [comment]
> > TIL, thanks! would it be possible to add some unit tests as well?
> 
> That sounds great, will issue another pr on that, still need some knowledge about how to `mock time`

Is it not possible to add directly to `TestManager`; I feel like we do not have enough unit tests; so having a test that go with PRs for the changes that is being would be awesome. Ultimately, the maintainers get to decide though.

**Comment by jonyhy96** [comment]
> Is it not possible to add directly to `TestManager`; I feel like we do not have enough unit tests; so having a test that go with PRs for the changes that is being would be awesome. Ultimately, the maintainers get to decide though.

Great, will try to add one. BTW do u have interest to add one?

**Comment by jonyhy96** [comment]
@manugupt1 Thanks for advice, unit test added.
PTAL @pacoxu @wzshiming

**Review by pohly — COMMENTED** [review]
Could it be that this PR is causing a data race in the unit test? Here's one example:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107677/pull-kubernetes-unit/1484567054118817792

I can't think of a reason why that PR should cause such a regression.

**Comment by jonyhy96** [comment]
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107677/pull-kubernetes-unit/1484567054118817792

Thanks! will take a look

**Pull request merged**